illumos-utils/src/zfs.rs

```diff
@@ -493,6 +506,196 @@ pub struct DatasetEnsureArgs<'a> {
+        .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?
+        .next()
+        .is_none())
+}
+
+fn ensure_mountpoint_empty(
+    path: &Utf8Path,
+) -> Result<(), EnsureDatasetErrorRaw> {
+    if is_directory_empty(path)? {
+        return Ok(());
+    }
+
+    let (Some(parent), Some(file)) = (path.parent(), path.file_name()) else {
+        return Err(EnsureDatasetErrorRaw::BadMountpoint(path.to_path_buf()));
+    };
+
+    // The directory is not empty. Let's make a new directory,
+    // with the "old-under-mountpoint-" prefix, and move all data there.
+
+    let prefix = format!("old-under-mountpoint-{file}");
+    let destination_dir = Utf8TempDir::with_prefix_in(prefix, parent)
+        .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?
+        .into_path();
+
+    let entries = path
+        .read_dir_utf8()
+        .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?;
+    for entry in entries {
+        let entry = entry
+            .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?;
+
+        // This would not work for renaming recursively, but we're only renaming
+        // a single directory's worth of files.
+        let src = entry.path();
+        let dst = destination_dir.as_path().join(entry.file_name());
+
+        std::fs::rename(src, dst)
+            .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?;
+    }
+
+    Ok(())
+}
+
+fn make_directory_immutable(
+    path: &Utf8Path,
+) -> Result<(), EnsureDatasetErrorRaw> {
+    let mut command = std::process::Command::new(PFEXEC);
+    let cmd = command.args(&["chmod", "S+ci", path.as_str()]);
+    execute(cmd).map_err(|err| EnsureDatasetErrorRaw::MakeImmutable(err))?;
+    Ok(())
+}
+
+fn make_directory_mutable(
+    path: &Utf8Path,
+) -> Result<(), EnsureDatasetErrorRaw> {
+    let mut command = std::process::Command::new(PFEXEC);
+    let cmd = command.args(&["chmod", "S-ci", path.as_str()]);
+    execute(cmd).map_err(|err| EnsureDatasetErrorRaw::MakeImmutable(err))?;
+    Ok(())
+}
+
+enum Immutability {
+    Yes,
+    No,
+    Unsupported,
+}
+
+impl Immutability {
+    fn as_immutable_as_filesystem_allows(&self) -> bool {
+        match self {
+            Immutability::Yes => true,
+            Immutability::No => false,
+            Immutability::Unsupported => true,
+        }
+    }
+
+    fn can_set_immutable(&self) -> bool {
+        match self {
+            Immutability::Yes => true,
+            Immutability::No => true,
+            Immutability::Unsupported => false,
+        }
+    }
+}
+
+fn is_directory_immutable(
+    path: &Utf8Path,
+) -> Result<Immutability, EnsureDatasetErrorRaw> {
+    let mut command = std::process::Command::new(PFEXEC);
+    let cmd = command.args(&["ls", "-d/v", path.as_str()]);
+    let output = execute(cmd)
+        .map_err(|err| EnsureDatasetErrorRaw::MakeImmutable(err))?;
+
+    // NOTE: It's probably worthwhile figuring out how to use stat for this
```

Comment:
good luck finding a portable API to do this though. A wrapper around C bindings might be necessary here; I don't think there are posix-y APIs to access this info.